drivers: modem: gsm: Some optimizations

ycsin opened the pull request:
Some optimizations for the `gsm_ppp` modem before the FSM implementation, most of them cherry picked from #38596

ycsin committed:
drivers: modem: gsm: Remove unused flags

The `setup_done` and `mux_setup_done` aren't being used
anywhere in the driver, therefore should be removed.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: modem_cmd_handler: Fix missing documentations

Fix missing documentations.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Customize error logs to ease debugging

Customize the error logs in each connection finalization steps
so that it is easier to trace the error.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Reorder gsm mux logs so that it will be printed

The gsm_configure will return if it fails to perform mux_enable,
therefore the disable part of the log will not be printed.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Remove unnecessary mux_enabled check

Remove redundant mux_enabled checks, the code execution will
not reach here if mux_enabled is false in the first place.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Go to next state if mux_attach successfully

Currently the mux_setup set the state to the next one after
uart_mux_alloc is successful even if the mux_setup fails which
can be a problem.

Set the state after both mux_setup & uart_mux_alloc are
successful.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Convert gsm_finalize_connection to work

Convert gsm_finalize_connection into a work so that the caller
work won't have to run again when gsm_finalize_connection
reschedule gsm_configure_work on error.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
cc @bbilas

rlubos reviewed (dismissed):
Looks reasonable

bbilas commented on drivers/modem/gsm_ppp.c:
```diff
@@ -697,12 +692,11 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 	}
 
-	/* Attached, clear retry counter */
+	/* Attached */
```
Is this comment necessary? IMO. `gsm->attached = true;` is much enough.

ycsin replied:
Agreed

ycsin committed:
drivers: modem: gsm: Init rssi_work_handle on gsm_init only

Initializing rssi_work_handle on gsm_start would (re)init it
unnecessarily everytime the gsm_start is invoked, we only need
to initialize it once in the gsm_init.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

bbilas commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1168,6 +1150,10 @@ static int gsm_init(const struct device *dev)
 			   K_PRIO_COOP(7), NULL);
 	k_thread_name_set(&gsm->workq.thread, "gsm_workq");
 
+#if defined(CONFIG_GSM_MUX)
```
Why not `if (IS_ENABLED(CONFIG_GSM_MUX)`?

ycsin replied:
I noticed that the `rssi_work_handle` is declared in the struct but is unused if `CONFIG_GSM_MUX=n`, if we exclude that then we have to leave the macro as is, otherwise we can convert quite a few `#if defined(CONFIG_GSM_MUX)` to `if (IS_ENABLED(CONFIG_GSM_MUX)`

ycsin replied:
I decided to keep this as-is for now. We will decide what to do with them in the next PR

bbilas commented:
There is still a couple of places where `#if defined` is mixed with `if (IS_ENABLED(CONFIG_X)` but IMO. it can be fixed later since smaller MRs are quite easier to review.

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -606,7 +605,7 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 
 	/* If attach check failed, we should not redo every setup step */
-	if (gsm->attach_retries) {
+	if (gsm->retries) {
```
Need to reset the `retries` somewhere so that this part of code can run normally when the modem is restarted.

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -598,32 +595,33 @@ static void rssi_handler(struct k_work *work)
+	struct gsm_modem *gsm = CONTAINER_OF(dwork, struct gsm_modem, gsm_configure_work);
 
 	/* If already attached, jump right to RSSI readout */
 	if (gsm->attached) {
```
Currently the driver set the `attached` flag to true [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/gsm_ppp.c#L701) and nothing resets it back to false. That means the second time the modem is started, A big part of the function will be [skipped](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/gsm_ppp.c#L606)?

bbilas replied:
It could be done in `gsm_ppp_stop` function for instance.

ycsin replied:
I think it's better to reset it inside `gsm_ppp_stop`, just make sure that it is cleared after `(void)k_sem_take(&gsm->sem_if_down, K_FOREVER);`

ycsin committed:
drivers: modem: gsm: Reset attached flag on stop

The 'attached' flag should be reset on gsm_ppp_stop, or else
some part of the gsm_finalize_connection won't be executed
during the next gsm_ppp_start.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: rssi_work should be submitted to gsm workq

The rssi_work_handle should be submitted to the gsm workqueue
using the gsm_work_reschedule, fix that.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: cancel rssi work only if GSM_MUX is enabled

There's no need to cancel rssi work when CONFIG_GSM_MUX isn't
enabled, since it is not scheduled.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: remove use of preprocessor directives for rssi

The rssi_work_handle always exists in the gsm struct, so use

```
if (IS_ENABLED(CONFIG_GSM_MUX))
```
instead of

```
#if defined(CONFIG_GSM_MUX)
```

for better code readability.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Define tx_lock timeout at the top

The tx_lock timeout is closely related to GSM_CMD_AT_TIMEOUT
& GSM_CMD_SETUP_TIMEOUT, and should be longer than both of
them, otherwise the gsm_ppp_stop might fail to lock the tx.

Define the timeout at the top of the driver and added a comment
to make it clear for the user.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
Ready for review.

jukkar reviewed (commented):
Overall looks good, but this PR is quite large and you added `drivers: modem: gsm: Update` which makes it even larger. I suggest that you submit that commit separately after this PR has been merged.

ycsin commented:
> you added `drivers: modem: gsm: Update` which makes it even larger

The commit that you are seeing is my local branch, it will not be included into this PR

rlubos commented on drivers/modem/modem_cmd_handler.h:
```diff
@@ -271,18 +277,25 @@ int modem_cmd_handler_init(struct modem_cmd_handler *handler,
+ *
+ * @param  *handler: command handler to unlock
+ *
+ * @retval None.
```
If the function does not return anything, the "retval"/"return" part of the docstring is typically skipped.

rerickson1 commented on drivers/modem/gsm_ppp.c:
```diff
@@ -618,10 +618,8 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
-			(void)gsm_work_reschedule(&gsm->gsm_configure_work,
-						K_SECONDS(1));
+			LOG_ERR("%s returned %d, %s", "AT", ret, "retrying...");
+			(void)gsm_work_reschedule(&gsm->gsm_configure_work, K_SECONDS(1));
```
Should there be a define for the` K_SECONDS(1)`? Its used a lot

ycsin replied:
fixed in new commit

ycsin committed:
drivers: modem: gsm: define retry delay in macro

Define the delay used to reschedule work when AT command failed
in a macro.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>
